### ISSUE TITLE
question: how the predict method is implemented in Neupy?

### ISSUE THREAD

**Issue by ivan-marroquin** [itdxer/neupy#190]
Hi !

I searched in the source code to understand how the predict method is implemented, without success. Can someone explain it to me?

In the attached script, I get the weights of the output layer after the model is trained (see lines 470 - 475). I gather the following information:
shape output layer weights: (8,1)
shape input data used to predict classes: (801,4) - To train the model, I used an array of (55,4)
shape output predicted classes: (801,1)

So, how does Neupy output an array of (801,1)?

Another question, the command line: 
print("output layer activation function {}\n".format(best_model.layers[2].weight))

gives this output: softplus-21/weight

I would like to know what is the meaning of 21? Different runs produce a new value.

Many thanks,

Ivan
[Script_and_data.zip](https://github.com/itdxer/neupy/files/1226536/Script_and_data.zip)

**Comment by itdxer** [comment]
Hi,

RE parameter names question, in neupy you can specify layer name, otherwise it would be generated automatically. The rule is simple, neupy takes class name and adds to it identifier that increments over the time

```python
In [6]: from neupy import layers

In [7]: layers.Input(3) > layers.Sigmoid(2) > layers.Sigmoid(4) > layers.Relu(1)
Out[7]: Input(3) > Sigmoid(2) > Sigmoid(4) > Relu(1)

In [8]: conn = _

In [9]: conn
Out[9]: Input(3) > Sigmoid(2) > Sigmoid(4) > Relu(1)

In [10]: for layer in conn:
    ...:     print(layer.name)
    ...:
input-1
sigmoid-1
sigmoid-2
relu-1

In [11]: conn2 = layers.Input(3) > layers.Sigmoid(2) > layers.Sigmoid(4) > layers.Relu(1)

In [12]: for layer in conn2:
    ...:     print(layer.name)
    ...:
input-2
sigmoid-3
sigmoid-4
relu-2
```

**Comment by itdxer** [comment]
Cannot give you a quick answer for your first question, will check the code and I will write back to you

**Comment by itdxer** [comment]
you output looks fine for me. `xtest_data` has shape `(801, 4)` then you propagate if through the hidden layer that has 8 hidden units and output output from the hidden layer should be `(801, 8)` and when you propagate through the final layer your shape is `(801, 1)` (because you multiplied it by matrix with shape `(8, 1)`). I hope it will help you

Also, The `Softmax` layer probably not the best choice for the hidden layer (line 252). I would suggest using `Tanh`, `Sigmoid` or `Relu`

**Comment by ivan-marroquin** [comment]
Thanks for the reply. Let me see if I understood well the way the predict method works:
- from the trained model, I get the weights of the hidden layer and output layer
- Matrix multiplication: Input (801,4) x Weights hidden layer (4,8) + bias => gives Output_1 (801,4)
- Matrix multiplication: Output_1 (801,4) x Weights output layer (8,1) + bias => gives Output_2 (801,1)

The final result is Output_2.

In the script that I sent you, I also use the bias for both hidden and output layers to build the neural network model. So in the matrix multiplication steps, the bias is just added to each output. Am I correct?

Also thanks for the suggestion to change the activation function for the hidden layer. With the data that I have at hand, I tried different combinations for the hidden/output layers. So far, the softmax produced the most consistent results. Once I have new data, I will keep  in mind your suggestion.

Many thanks,

Ivan

**Comment by itdxer** [comment]
Output_1 will have shape (801, 8), because (801,4) x (4,8)

> So in the matrix multiplication steps, the bias is just added to each output. Am I correct?

yes

**Comment by ivan-marroquin** [comment]
Many thanks for helping me to understand the predict method!
